src/OledDisplay.cpp

```diff
@@ -0,0 +1,195 @@
+{
+    // get the width and height of the display
+    int maxWidth = display->getWidth();
+    int maxHeight = display->getHeight();
+
+    // pxZittern in +x (0 - 6 px)
+    int ex = 2 * (mExtra % 4);
+
+    // set the font size based on the display size
+    switch (line) {
+    case 1:
+        if (maxWidth > 120 && maxHeight > 60)
+            display->setFont(u8g2_font_ncenB14_tr); // large display
+        else
+            display->setFont(u8g2_font_logisoso16_tr); // small display
+        break;
+    case 4:
+        if (maxWidth > 120 && maxHeight > 60)
+            display->setFont(u8g2_font_5x8_tr); // large display
+        else
+            display->setFont(u8g2_font_5x8_tr); // small display
+        break;
+    default:
+        if (maxWidth > 120 && maxHeight > 60)
+            display->setFont(u8g2_font_ncenB10_tr); // large display
+        else
+            display->setFont(u8g2_font_5x8_tr); // small display
+        break;
+    }
+
+    // get the font height
+    int fontHeight = display->getMaxCharHeight();
+
+    // calculate the starting position for the text
+    if (line == 1)
+        dispX = 25 + ex;
+    else
+        dispX = 5 + ex;
+
+    dispY += fontHeight + 1;
+
+    display->drawStr(dispX, dispY, text.c_str());
+
+    // Logo und Blitz
+    display->drawXBMP(10 + ex, 1, 8, 17, bmp_arrow);
+    display->drawXBM(maxWidth - 24 + ex, 2, 16, 16, bmp_logo);
+}
+
+void DisplayClass::loop()
+{
+    if (display_type == 0)
+        return;
+
+    if ((millis() - _lastPublish) > period) {
+        float totalPower = 0;
+        float totalYieldDay = 0;
+        float totalYieldTotal = 0;
+
+        display->clear();
+        display->firstPage();
+
+        do {
+            // actual seconds
+            unsigned long currentMillis = millis();
+
+            // set Contrast of the Display to raise the lifetime
+            display->setContrast(60);
+            int isprod=0;
+
+            for (uint8_t i = 0; i < Hoymiles.getNumInverters(); i++) {
+                auto inv = Hoymiles.getInverterByPos(i);
+                if (inv == nullptr)
+                    continue;
+                
+                if(inv->isProducing())
+                    isprod++;
+
+                totalPower += inv->Statistics()->getChannelFieldValue(CH0, FLD_PAC);
+                totalYieldDay += inv->Statistics()->getChannelFieldValue(CH0, FLD_YD);
+                totalYieldTotal += inv->Statistics()->getChannelFieldValue(CH0, FLD_YT);
+            }
+
+            //=====> Actual Production
+            if ((totalPower > 0) && (isprod > 0)) {
+                display->setPowerSave(false);
+                if (totalPower > 999) {
+                    sprintf(fmtText, "%2.1f", (totalPower / 1000));
+                    printText(String(fmtText) + F(" kW"), 1);
+                } else {
+                    sprintf(fmtText, "%3.0f", totalPower);
+                    printText(String(fmtText) + F(" W"), 1);
```

Comment:
Das " W" kann in den Format String eine Zeile weiter oben mit hinein. Dadurch spart man sich einen Constructor aufruf der String Klasse.